packages/compass-crud/src/stores/crud-store.ts

```diff
@@ -1584,44 +1570,104 @@ type ErrorOrResult =
-  ) => Promise<ErrorOrResult>
+  query: Partial<QueryState>,
+  object: unknown,
+  modificationType: 'update' | 'replace'
 ): Promise<ErrorOrResult> {
-  const opts = { returnDocument: 'after', promoteValues: false } as const;
+  let options: { returnDocument: 'before' | 'after'; promoteValues: false } = {
+    returnDocument: 'after',
+    promoteValues: false,
+  };
+  let error: (Error & { codeName?: string; code?: any }) | undefined;
+  let document;
 
-  let [error, d] = await doFindAndModify(ds, ns, opts);
-  const originalError = error;
+  try {
+    if (modificationType === 'update') {
+      document = await ds.findOneAndUpdate(
+        ns,
+        query,
+        object as Document,
+        options
+      );
+    } else {
+      document = await ds.findOneAndReplace(
+        ns,
+        query,
+        object as Document,
+        options
+      );
+    }
+  } catch (e) {
+    error = e as Error;
+  }
 
-  // 6371402 is "'findAndModify with encryption only supports new: false'"
-  if (error && +(error?.code ?? 0) === 63714_02) {
-    // For encrypted documents, returnDocument: 'after' is unsupported on the server
-    const fallbackOpts = {
-      returnDocument: 'before',
-      promoteValues: false,
-    } as const;
-    [error, d] = await doFindAndModify(ds, ns, fallbackOpts);
+  if (!error) {
+    return [undefined, document] as ErrorOrResult;
+  }
 
-    if (!error) {
-      let docs;
+  if (error.codeName === 'ShardKeyNotFound') {
+    try {
+      const docs = await ds.find(ns, query, options);
+      [document] = docs;
+    } catch (e) {
+      return [e, undefined] as ErrorOrResult;
+    }
+
+    if (document) {
       try {
-        docs = await ds.find(ns, { _id: d!._id } as any, fallbackOpts);
+        if (modificationType === 'update') {
+          await ds.updateOne(ns, { _id: document._id }, object as Document);
+        } else {
+          await ds.replaceOne(ns, { _id: document._id }, object as Document);
+        }
       } catch (e) {
-        error = e as Error;
+        return [e, undefined] as ErrorOrResult;
       }
-      if (error || !docs || !docs.length) {
-        // Race condition -- most likely, somebody else
-        // deleted the document between the findAndModify command
-        // and the find command. Just return the original error.
-        error = originalError;
-        d = undefined;
+    }
+  }
+
+  // 6371402 is "'findAndModify with encryption only supports new: false'"
+  if (+(error?.code ?? 0) === 63714_02) {
+    options = {
+      returnDocument: 'before',
+      promoteValues: false,
+    };
+    try {
+      if (modificationType === 'update') {
+        document = await ds.findOneAndUpdate(
+          ns,
+          query,
+          object as Document,
+          options
+        );
       } else {
-        [d] = docs;
+        document = await ds.findOneAndReplace(
+          ns,
+          query,
+          object as Document,
+          options
+        );
```

Comment:
There's quite a bit of duplication now, this code block shows up 4 times in the same function essentially... I'd either store the arguments as an array and then simplify these calls to `await ds.findOneAndReplace(...args);`, or dynamically select the method (`const method = modificationType === 'update' ? 'findOneAndUpdate' : 'findOneAndReplace'; await ds[method](ns, query, object, options);`)